xarray/core/indexes.py

```diff
@@ -1,12 +1,207 @@
+
+    @classmethod
+    def from_variables(
+        cls, variables: Dict[Hashable, "Variable"], **kwargs
+    ):  # pragma: no cover
+        raise NotImplementedError()
+
+    def to_pandas_index(self) -> pd.Index:
+        """Cast this xarray index to a pandas.Index object or raise a TypeError
+        if this is not supported.
+
+        This method is used by all xarray operations that expect/require a
+        pandas.Index object.
+
+        """
+        raise TypeError(f"{type(self)} cannot be cast to a pandas.Index object.")
+
+    def equals(self, other):  # pragma: no cover
+        raise NotImplementedError()
+
+    def union(self, other):  # pragma: no cover
+        raise NotImplementedError()
+
+    def intersection(self, other):  # pragma: no cover
+        raise NotImplementedError()
+
+
+class PandasIndexAdapter(Index, ExplicitlyIndexedNDArrayMixin):
+    """Wrap a pandas.Index to preserve dtypes and handle explicit indexing."""
+
+    __slots__ = ("array", "_dtype")
+
+    def __init__(
+        self, array: Any, dtype: DTypeLike = None, coord_name: Optional[Hashable] = None
+    ):
+        if coord_name is None:
+            coord_name = tuple()
+        super().__init__(coord_name)
+
+        self.array = utils.safe_cast_to_index(array)
+
+        if dtype is None:
+            if isinstance(array, pd.PeriodIndex):
+                dtype_ = np.dtype("O")
+            elif hasattr(array, "categories"):
+                # category isn't a real numpy dtype
+                dtype_ = array.categories.dtype
+            elif not utils.is_valid_numpy_dtype(array.dtype):
+                dtype_ = np.dtype("O")
+            else:
+                dtype_ = array.dtype
+        else:
+            dtype_ = np.dtype(dtype)
+        self._dtype = dtype_
+
+    @classmethod
+    def from_variables(cls, variables: Dict[Hashable, "Variable"], **kwargs):
+        if len(variables) > 1:
+            raise ValueError("Cannot set a pandas.Index from more than one variable")
+
+        varname, var = list(variables.items())[0]
+        return cls(var.data, dtype=var.dtype, coord_name=varname)
+
+    def to_pandas_index(self) -> pd.Index:
+        return self.array
+
+    @property
+    def dtype(self) -> np.dtype:
+        return self._dtype
+
+    def __array__(self, dtype: DTypeLike = None) -> np.ndarray:
+        if dtype is None:
+            dtype = self.dtype
+        array = self.array
+        if isinstance(array, pd.PeriodIndex):
+            with suppress(AttributeError):
+                # this might not be public API
+                array = array.astype("object")
+        return np.asarray(array.values, dtype=dtype)
+
+    @property
+    def shape(self) -> Tuple[int]:
+        return (len(self.array),)
+
+    def equals(self, other):
+        if isinstance(other, pd.Index):
+            return self.array.equals(other)
+        else:
+            return self.array.equals(other.array)
```

Comment:
nit: I would suggest re-writing these checks in the form of a "normalization" step. This makes it clearer that the "normal" value for `other` is a `PandasIndex` object, and avoids duplication, e.g.,
```suggestion
        if isinstance(other, pd.Index):
            other = PandasIndex(other)
        return isinstance(other, PandasIndex) and self.array.equals(other.array)
```